caikit/core/toolkit/destroyable_process.py

```diff
@@ -0,0 +1,193 @@
+            log.error(
+                "<COR12037430E>",
+                "get_or_throw called on process, but it has not finished running",
+            )
+
+        # Update the result and throw if it's an error
+        error = self.error
+        if error is not None:
+            raise error
+        return self.__result
+
+    def destroy(self):
+        """Cancel any in-progress work"""
+        self.__destroyed = True
+        if self.is_alive() or not self.__started:
+            self.__canceled = True
+        if self.__started:
+            self.terminate()
+            self.join(self._destroy_grace_period)
+            self.kill()
+            self.join()
+            self._completion_event.set()
+
+    ## Process Interface ##
+
+    def start(self):
+        if self.destroyed:
+            err_msg = "Not starting work on pre-canceled process"
+            log.warning("<COR42191929W>", err_msg)
+            self._completion_event.set()
+            self.__result = RuntimeError(err_msg)
+            return
+        self.__started = True
+        return super().start()
+
+    def join(self, *args, **kwargs):
+        if self.destroyed and not self.__started:
+            return
+        return super().join(*args, **kwargs)
+
+    # NOTE: This functionality is covered by the unit tests, but pytest-cov is
+    #   not correctly collecting the coverage info since it executes inside the
+    #   subprocess
+    def run(self):  # pragma: no cover
+        try:
+            # Run and indicate to the parent that no
+            super().run()
+
+        # Catch any errors thrown within a subprocess so that they can be
+        # forwarded to the parent
+        # pylint: disable=broad-exception-caught
+        except Exception as err:
+            err_str = repr(err)
+            log.error(
+                "<COR69863806E>",
+                "Caught exception in destroyable process: %s",
+                err_str,
+                exc_info=True,
+            )
+            self._child_conn.send(err)
+        finally:
+            self._completion_event.set()
+
+    ## Impl ##
+
+    def _update_result(self):
+        if self._parent_conn.poll():
+            self.__result = self._parent_conn.recv()
+
+    @property
+    def error(self) -> Optional[Exception]:
+        self._update_result()
+        if isinstance(self.__result, Exception):
+            return self.__result
+
+    @property
+    def completion_event(self) -> multiprocessing.Event:
+        return self._completion_event
+
+    # NOTE: This functionality is covered by the unit tests, but pytest-cov is
+    #   not correctly collecting the coverage info since it executes inside the
+    #   subprocess
+    def _target_wrapper(self, target, args, kwargs):  # pragma: no cover
+        result = target(*(args or []), **(kwargs or {}))
+        log.debug3("Process target result: %s", result)
+        if self._return_result:
+            self._child_conn.send(result)
```

Comment:
can this create issue since size of the `result` (pickled instance of the trained class) here can be bigger and pipe would have limitation on the size of data transfer, so either it can take quite a while and meanwhile `join` and `close` can create deadlocks ?

Reply:
Ah, yeah, this is actually explicitly why I added the `return_results` flag: When this class is used for training, [it will NOT return the result at all](https://github.com/caikit/caikit/pull/310/files#diff-01866f980d3c84ae181021e96cf0ccd560747994902c657fb5fad90e3edde31fR104). I added this whole result-returning capability to make this a more generic tool, similar to `DestroyableThread`. Technically, this capability is not immediately needed, but I liked the symmetry and I think it likely will be useful down the road (possibly for inference subprocess wrapping).

Reply:
oh gotcha. Make sense. Thanks!